pkg/controller/cloneset/sync/cloneset_sync_utils.go

```diff
@@ -0,0 +1,196 @@
+/*
+Copyright 2021 The Kruise Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sync
+
+import (
+	appspub "github.com/openkruise/kruise/apis/apps/pub"
+	appsv1alpha1 "github.com/openkruise/kruise/apis/apps/v1alpha1"
+	clonesetcore "github.com/openkruise/kruise/pkg/controller/cloneset/core"
+	clonesetutils "github.com/openkruise/kruise/pkg/controller/cloneset/utils"
+	"github.com/openkruise/kruise/pkg/util"
+	"github.com/openkruise/kruise/pkg/util/lifecycle"
+	"github.com/openkruise/kruise/pkg/util/specifieddelete"
+	v1 "k8s.io/api/core/v1"
+	intstrutil "k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/klog"
+	"k8s.io/utils/integer"
+)
+
+type expectationDiffs struct {
+	// for scale
+	scaleDiff            int
+	scaleOldRevisionDiff int
+	deleteReadyLimit     int
+	useSurge             int
+	useOldRevisionSurge  int
+	// for update
+	updateDiff           int
+	updateMaxUnavailable int
+}
+
+// This is the most important algorithm in cloneset-controller.
+// It calculates the pod numbers to scaling and updating for current CloneSet.
+func calculateDiffsWithExpectation(cs *appsv1alpha1.CloneSet, pods []*v1.Pod, updateRevision string) (res expectationDiffs) {
```

Comment:
I suggest to have a separate document to explain the algorithm. I think code comments are not enough to explain the intentions well. When user observed behaviors that are out of their expectations, they need to have a reference to justify the observations.